// @pm2/io/build/main/src/wrapper/httpWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const proxy_1 = require("../utils/proxy");
class HttpWrapper {
    constructor(metricFeature) {
        this.metricFeature = metricFeature;
    }
    init(opts, http) {
        let glMeter;
        let glLatency;
        if (!opts.name)
            opts.name = 'http';
        glMeter = this.metricFeature.meter({
            name: opts.name.toUpperCase(),
            samples: 60,
            unit: 'req/min'
        });
        glLatency = this.metricFeature.histogram({
            measurement: 'mean',
            name: `pmx:${opts.name}:latency`,
            unit: 'ms'
        });
        const ignoreRoutes = function (url) {
            for (let i = 0; i < opts.ignore_routes.length; ++i) {
                if (url.match(opts.ignore_routes[i]) !== null) {
                    return true;
                }
            }
            return false;
        };
        proxy_1.default.wrap(http.Server.prototype, ['on', 'addListener'], function (addListener) {
            return function (event, listener) {
                const self = this;
                const overloadedFunction = function (request, response) {
                    glMeter.mark();
                    let httpStart = {
                        url: request.url,
                        start: Date.now()
                    };
                    response.once('finish', function () {
                        if (!ignoreRoutes(httpStart.url)) {
                            glLatency.update(Date.now() - httpStart.start);
                        }
                    });
                };
                if (!(event === 'request' && typeof listener === 'function')) {
                    return addListener.apply(self, arguments);
                }
                if (self.__overloaded !== true) {
                    self.on('removeListener', function onRemoveListener() {
                        setTimeout(function () {
                            if (self.listeners('request').length === 1) {
                                self.removeListener('request', overloadedFunction);
                                self.removeListener('removeListener', onRemoveListener);
                                self.__overloaded = false;
                            }
                        }, 200);
                    });
                    addListener.call(self, event, overloadedFunction);
                    self.__overloaded = true;
                }
                return addListener.apply(self, arguments);
            };
        });
        return http;
    }
}
exports.default = HttpWrapper;
