// @pm2/io/build/main/src/features/notify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util = require("util");
const semver = require("semver");
const json_1 = require("../utils/json");
const configuration_1 = require("../configuration");
const transport_1 = require("../utils/transport");
const debug_1 = require("debug");
const debug = debug_1.default('axm:notify');
class NotifyOptions {
}
exports.NotifyOptions = NotifyOptions;
exports.NotifyOptionsDefault = {
    level: 'info',
    catchExceptions: true
};
class NotifyFeature {
    constructor() {
        this.options = exports.NotifyOptionsDefault;
        this.levels = ['fatal', 'error', 'warn', 'info', 'debug', 'trace'];
    }
    init(options) {
        if (options) {
            this.options = options;
        }
        if (this.options && this.options.catchExceptions) {
            if (process.env.CATCH_CONTEXT_ON_ERROR === 'true' && (semver.satisfies(process.version, '< 8.0.0') ||
                (semver.satisfies(process.version, '< 10.0.0') && !process.env.FORCE_INSPECTOR))) {
                debug(`Inspector is not available on node version ${process.version} !`);
            }
            if (process.env.CATCH_CONTEXT_ON_ERROR === 'true' && semver.satisfies(process.version, '>= 10.0.0') ||
                (semver.satisfies(process.version, '>= 8.0.0') && process.env.FORCE_INSPECTOR)) {
                const NotifyInspector = require('./notifyInspector').default;
                NotifyInspector.catchAllDebugger();
            }
            else {
                this.catchAll();
            }
        }
        return {
            notifyError: this.notifyError
        };
    }
    notifyError(err, level) {
        if (!(err instanceof Error)) {
            console.trace('[PM2-IO-APM] You should use notify with an Error object');
            return -1;
        }
        if (!level || this.levels.indexOf(level) === -1) {
            return transport_1.default.send({
                type: 'process:exception',
                data: json_1.default.jsonize(err)
            });
        }
        if (this.levels.indexOf(this.options.level) >= this.levels.indexOf(level)) {
            return transport_1.default.send({
                type: 'process:exception',
                data: json_1.default.jsonize(err)
            });
        }
        return null;
    }
    catchAll(opts) {
        if (opts === undefined) {
            opts = { errors: true };
        }
        configuration_1.default.configureModule({
            error: opts.errors
        });
        if (process.env.exec_mode === 'cluster_mode') {
            return false;
        }
        const self = this;
        function getUncaughtExceptionListener(listener) {
            return function uncaughtListener(err) {
                let error = err && err.stack ? err.stack : err;
                if (err && err.length) {
                    err._length = err.length;
                    delete err.length;
                }
                if (listener === 'unhandledRejection') {
                    console.log('You have triggered an unhandledRejection, you may have forgotten to catch a Promise rejection:');
                }
                console.error(error);
                let errObj;
                if (err) {
                    errObj = self._interpretError(err);
                }
                transport_1.default.send({
                    type: 'process:exception',
                    data: errObj !== undefined ? errObj : { message: 'No error but ' + listener + ' was caught!' }
                });
                if (!process.listeners(listener).filter(function (listener) {
                    return listener !== uncaughtListener;
                }).length) {
                    if (listener === 'uncaughtException') {
                        process.exit(1);
                    }
                }
            };
        }
        if (opts.errors === true && util.inspect(process.listeners('uncaughtException')).length === 2) {
            process.once('uncaughtException', getUncaughtExceptionListener('uncaughtException'));
            process.once('unhandledRejection', getUncaughtExceptionListener('unhandledRejection'));
        }
        else if (opts.errors === false
            && util.inspect(process.listeners('uncaughtException')).length !== 2) {
            process.removeAllListeners('uncaughtException');
            process.removeAllListeners('unhandledRejection');
        }
    }
    expressErrorHandler() {
        configuration_1.default.configureModule({
            error: true
        });
        return function errorHandler(err, req, res, next) {
            if (res.statusCode < 400)
                res.statusCode = 500;
            err.url = req.url;
            err.component = req.url;
            err.action = req.method;
            err.params = req.body;
            err.session = req.session;
            transport_1.default.send({
                type: 'process:exception',
                data: json_1.default.jsonize(err)
            });
            return next(err);
        };
    }
    _interpretError(err) {
        let sErr = {
            message: null,
            stack: null
        };
        if (err instanceof Error) {
            // Error object type processing
            sErr = err;
        }
        else {
            // JSON processing
            sErr.message = err;
            sErr.stack = err;
        }
        return json_1.default.jsonize(sErr);
    }
}
exports.NotifyFeature = NotifyFeature;
